pushsync/simulation_test.go

```diff
@@ -0,0 +1,292 @@
+)
+
+// test syncer using pss
+// the test
+// * creates a simulation with connectivity loaded from a snapshot
+// * for each test case, two nodes are chosen randomly, an uploader and a downloader
+// * uploader uploads a number of chunks
+// * wait until the uploaded chunks are synced
+// * downloader downloads the chunk
+// Testcasrs are run concurrently
+func TestPushSyncSimulation(t *testing.T) {
+	nodeCnt := *nodeCntFlag
+	chunkCnt := *chunkCntFlag
+	testcases := *testCasesFlag
+
+	err := testSyncerWithPubSub(nodeCnt, chunkCnt, testcases, newServiceFunc)
+	if err != nil {
+		t.Fatal(err)
+	}
+}
+
+func testSyncerWithPubSub(nodeCnt, chunkCnt, testcases int, sf simulation.ServiceFunc) error {
+	sim := simulation.NewInProc(map[string]simulation.ServiceFunc{
+		"pushsync": sf,
+	})
+	defer sim.Close()
+
+	ctx, cancel := context.WithTimeout(context.Background(), 60*time.Second)
+	defer cancel()
+	err := sim.UploadSnapshot(ctx, fmt.Sprintf("../network/stream/testing/snapshot_%d.json", nodeCnt))
+	if err != nil {
+		return err
+	}
+
+	log.Debug("Snapshot loaded")
+	result := sim.Run(ctx, func(ctx context.Context, sim *simulation.Simulation) error {
+		errc := make(chan error)
+		for j := 0; j < testcases; j++ {
+			j := j
+			go func() {
+				err := uploadAndDownload(ctx, sim, nodeCnt, chunkCnt, j)
+				select {
+				case errc <- err:
+				case <-ctx.Done():
+				}
+			}()
+		}
+		i := 0
+		for err := range errc {
+			if err != nil {
+				return err
+			}
+			i++
+			if i >= testcases {
+				return nil
+			}
+		}
+		return nil
+	})
+
+	if result.Error != nil {
+		return fmt.Errorf("simulation error: %v", result.Error)
+	}
+	return nil
+}
+
+// pickNodes selects 2 distinct
+func pickNodes(n int) (i, j int) {
+	i = rand.Intn(n)
+	j = rand.Intn(n - 1)
+	if j >= i {
+		j++
+	}
+	return
+}
+
+func uploadAndDownload(ctx context.Context, sim *simulation.Simulation, nodeCnt, chunkCnt, i int) error {
+	// chose 2 random nodes as uploader and downloader
+	u, d := pickNodes(nodeCnt)
+	// setup uploader node
+	uid := sim.UpNodeIDs()[u]
+	p := sim.MustNodeItem(uid, bucketKeyPushSyncer).(*Pusher)
+	// setup downloader node
+	did := sim.UpNodeIDs()[d]
+	// the created tag indicates the uploader and downloader nodes
+	tagname := fmt.Sprintf("tag-%v-%v-%d", label(uid[:]), label(did[:]), i)
+	log.Debug("uploading", "peer", uid, "chunks", chunkCnt, "tagname", tagname)
+	tag, what, err := upload(ctx, p.store.(*localstore.DB), p.tags, tagname, chunkCnt)
+	if err != nil {
+		return err
+	}
+	log.Debug("uploaded", "peer", uid, "chunks", chunkCnt, "tagname", tagname)
+
+	// wait till pushsync is done
+	syncTimeout := 30 * time.Second
+	sctx, cancel := context.WithTimeout(ctx, syncTimeout)
+	err = tag.WaitTillDone(sctx, chunk.StateSynced)
+	if err != nil {
+		log.Debug("tag", "tag", tag)
+		cancel()
+		return fmt.Errorf("error waiting syncing: %v", err)
+	}
+	cancel()
```

Comment:
why do you need to call this here? `WaitTillDone` already returned. you could probably make do with just a `defer` statement in L152 and remove L155

Reply:
well... why need it at all? the call is anyway symbolic here (if `WaitTillDone` returned - you either succeeded or there was a context timeout). just use one `defer` statement and call `cancel` once instead of two different places